Docs/Configuration.tex

```diff
@@ -4422,6 +4413,120 @@ \subsection{Security Properties}\label{miscsecurityprops}
 
 \end{enumerate}
 
+\subsection{Serial Properties}\label{miscserialprops}
+
+\begin{enumerate}
+\item
+  \texttt{BaudRate}\\
+  \textbf{Type}: \texttt{plist\ integer}\\
+  \textbf{Failsafe}: \texttt{115200}\\
+  \textbf{Description}: Set the baud rate for serial port.
+
+  This option will override the value of \texttt{gEfiMdeModulePkgTokenSpaceGuid.PcdSerialBaudRate}
+  defined in \href{https://github.com/acidanthera/audk/blob/master/MdeModulePkg/MdeModulePkg.dec}{MdeModulePkg.dec}.
+
+\item
+  \texttt{ClockRate}\\
+  \textbf{Type}: \texttt{plist\ integer}\\
+  \textbf{Failsafe}: \texttt{1843200}\\
+  \textbf{Description}: Set the clock rate for serial port.
+
+  This option will override the value of \texttt{gEfiMdeModulePkgTokenSpaceGuid.PcdSerialClockRate}
+  defined in \href{https://github.com/acidanthera/audk/blob/master/MdeModulePkg/MdeModulePkg.dec}{MdeModulePkg.dec}.
+
+\item
+  \texttt{ExtendedTxFifoSize}\\
+  \textbf{Type}: \texttt{plist\ integer}\\
+  \textbf{Failsafe}: \texttt{64}\\
+  \textbf{Description}: Set the extended transmit FIFO size for serial port.
+
+  This option will override the value of \texttt{gEfiMdeModulePkgTokenSpaceGuid.PcdSerialExtendedTxFifoSize}
+  defined in \href{https://github.com/acidanthera/audk/blob/master/MdeModulePkg/MdeModulePkg.dec}{MdeModulePkg.dec}.
+
+\item
+  \texttt{FifoControl}\\
+  \textbf{Type}: \texttt{plist\ integer}\\
+  \textbf{Failsafe}: \texttt{0x07}\\
+  \textbf{Description}: Configure serial port FIFO Control settings.
+
+  This option will override the value of \texttt{gEfiMdeModulePkgTokenSpaceGuid.PcdSerialFifoControl}
+  defined in \href{https://github.com/acidanthera/audk/blob/master/MdeModulePkg/MdeModulePkg.dec}{MdeModulePkg.dec}.
+
+\item
+  \texttt{Init}\\
+  \textbf{Type}: \texttt{plist\ boolean}\\
+  \textbf{Failsafe}: \texttt{false}\\
+  \textbf{Description}: Perform serial port initialisation.
+  When this option is set to \texttt{false}, no keys from this section will be overridden.
+
+  This option will perform serial port initialisation within OpenCore prior to enabling
+  (any) debug logging.
+
+  Refer to the \hyperref[troubleshootingdebug]{\texttt{Debugging}} section for details.
+
+\item
+  \texttt{LineControl}\\
+  \textbf{Type}: \texttt{plist\ integer}\\
+  \textbf{Failsafe}: \texttt{0x07}\\
+  \textbf{Description}: Configure serial port Line Control settings.
+
+  This option will override the value of \texttt{gEfiMdeModulePkgTokenSpaceGuid.PcdSerialLineControl}
+  defined in \href{https://github.com/acidanthera/audk/blob/master/MdeModulePkg/MdeModulePkg.dec}{MdeModulePkg.dec}.
+
+\item
+  \texttt{PciDeviceInfo}\\
+  \textbf{Type}: \texttt{plist\ data}\\
+  \textbf{Failsafe}: \texttt{0xFF}\\
+  \textbf{Description}: Set PCI serial device information.
```

Comment:
@joevt Would you please provide more information on how to convert a device path to the formatted PciDeviceInfo bytes? Thanks!

Reply:
My current theory is that I just need the device number and function number from each PCI device in the UEFI device path for the first two bytes of each node in the PciDeviceInfo path. The second two bytes I'll leave as 00 00. The last node is followed by an FF.

The following code find PCIe serial ports and outputs their paths.
```
ioreg -p IODeviceTree -lw0 | perl -e '
	$ioregpath=""; $pcipath=""; while (<>) {
		if ( /^([ |]*)\+\-o (.+)  <class (\w+)/ ) {
			$indent = (length $1) / 2; $name = $2; $class = $3;
			$ioregpath =~ s|^((/[^/]*){$indent}).*|$1/$name|;
			$pcipath =~ s|^((/[^/]*){$indent}).*|$1/|;
			$name = ""; $uid = 0; $classcode = "";
		}
		   if ( $class eq "IOACPIPlatformDevice" ) { if ( /^[ |]*"name" = <"([^\"]*)">/ ) { $name = $1; }; if ( /^[ |]*"_UID" = "(\d+)"/ ) { $uid = $1; } }
		elsif ( $class eq "IOPCIDevice" ) { if ( /^[ |]*"pcidebug" = "\d+:(\d+):(\d+).*"/ ) { $name = sprintf("Pci(0x%x,0x%x)", $1, $2);  } if ( /^[ |]*"class-code" = <(\w+)>/ ) { $classcode = $1; } }
		if ( /^[ |]*}/ && $name ) {
			if ( $class eq "IOACPIPlatformDevice" ) {
				   if ($name eq "PNP0A03") { $name = sprintf ("PciRoot(0x%x)", $uid); }
				elsif ($name eq "PNP0A08") { $name = sprintf ("PciRoot(0x%x)", $uid); }
				elsif ($name =~ /PNP..../) { $name = sprintf ("Acpi(%s,0x%x)", $name, $uid); }
				# not translating all ACPI types since we only care about PCI devices
			}
			$pcipath .= $name;
			if ( $classcode eq "02000700" ) {
				$serialdevicepath = $pcipath =~ s/.*PciRoot\(0x0\)//r =~ s|/Pci\(||gr =~ s|\)| 00 00 |gr =~ s|0x||gr =~ s|,| |gr =~ s|\b(\w)\b|0\1|gr =~ s|$|FF|r;
				print $ioregpath =~ s|/Root/||r . "\n";
				print $pcipath =~ s|/*||r  . "\n"; 
				print "xxd -p -r <<< \"" . $serialdevicepath . "\" | base64\n";
				print "\n";
			}
		}
	}
'
```

This is the result:
```
/PCI0@0/P0P9@9/P9P2@0/P2P4@1/PXS3@0
PciRoot(0x0)/Pci(0x9,0x0)/Pci(0x0,0x0)/Pci(0x1,0x0)/Pci(0x0,0x0)
xxd -p -r <<< "09 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 FF" | base64

/PCI0@0/P0P9@9/P9P2@0/P2P4@1/PXS3@0,2
PciRoot(0x0)/Pci(0x9,0x0)/Pci(0x0,0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)
xxd -p -r <<< "09 00 00 00 00 00 00 00 01 00 00 00 00 02 00 00 FF" | base64
```


The following can take a UEFI device path and produce the serial PciDevicePath:
```
serialpath=PciRoot(0x0)/Pci(0x9,0x0)/Pci(0x0,0x0)/Pci(0x1,0x0)/Pci(0x0,0x0)
perl -pe 's/PciRoot(0x0)//;s|/Pci\(||g;s|\)| 00 00 |g;s|0x||g;s|,| |g;s|\b(\w)\b|0\1|g; s|$|FF|' <<< "$serialpath" | xxd -p -r | base64
```

I'll let you know when I find out if this works.